conda crawler implementation

lamarrr opened the pull request:
closes #535

This Merge request is intended to track the work in implementing the crawler for conda source packages.

## Background

Conda exposes packages in a different format from other Python repositories like pypi. Conda is a Python environment locked to a specific Python version.
Conda deals with packages locked to a specific version for a version of the channel, this ensures packages do not break due to one incompatibility or another as the packages are managed for compatibility, similar to how you'd ship a docker container.
The primary consumption point is the "packages" themselves which are accompanied with scripts to modify the environment and setup the packages and dependencies which are then consumed by the setup application, the packages may also contain DLLs, scripts, compiled Python binary (.pyc), python code. etc.
The structure of conda repositories and their indexing process is described here: https://docs.conda.io/projects/conda-build/en/stable/concepts/generating-index.html

Conda has three main channels: anaconda-main, anaconda-r, and conda-forge which is more geared toward business uses

We crawl both the packages and the source code (not always specified) for the licensing metadata and other metadata about the package.

the source from which the conda packages are created is often but not always provided via a URL that links a compressed source file hosted externally, **sometimes** via GitHub, or another website. note that this is a file and not a git repository.
the main conda package is hosted on the conda channels themselves and is compressed and contains necessary licensing information, compilers, environment configuration scripts, dependencies, etc. that are needed to make the package work.

The crawler uses the coordinates of the syntax: 

```yaml
type: conda | condasource
provider: conda-forge | anaconda-main | anaconda-r
namespace: ${architecture}
name: any
revision:  (${version} |  )-(${buildversion} |  )
```
i.e. 
```txt
conda/conda-forge/linux-aarch64/numpy/1.13.0
condasource/conda-forge/linux-aarch64/numpy/1.13.0
conda/conda-forge/-/numpy/1.13.0/
conda/conda-forge/linux-aarch64/numpy/-py36
```

where 
type (required): conda or condasource
namespace (optional): architecture and OS of the package to be crawled i.e. win64, linux-aarch64,  if no architecture is specified any architecture is chosen.
package name: name of the package
provider (required): channel on which the package will be crawled. conda-forge, anaconda-main, or anaconda-r
revision (optional): package version and optional build version i.e. 0.3.0, 0.3.0-py36hffe2fc. if it is a conda coordinate type and the build version of the package is usually a conda-specific representation of the build tools and environment configuration, and build iteration of the package. i.e. for a Python 3.9 environment, this could be py39H443E.
 if none is specified, the latest one will be selected using the package's timestamp.


Conda-forge is a community effort and packages are published by opening PRs on their GitHub repository as described here https://conda-forge.org/docs/maintainer/adding_pkgs.html

qtomlinson commented:
It is exciting to see that a new harvester is being implemented!  This pull request provides a solid foundation for future enhancements.  A discussion is needed on the proposed coordinates, e.g. conda/conda-forge/-/numpy/1.13.0_linux-aarch64/py36. In the above proposal, toolVersion is used to represent the build (string) in conda-forge.  Points to consider:
- toolVersion is already used internally by ClearlyDefined as the harvest tool versioning, e.g.  Using spec.toolVersion will cause conflicts.  
- Adding /py36 adds complexity to [service APIs](https://api.clearlydefined.io/api-docs/#/).  For instance, harvest data api expects
/harvest/{type}/{provider}/{namespace}/{name}/{revision}/{tool}.
- Concatenating version and architecture with `_` does not handle versions that contain `_`, e.g. "version": "1.30.0_2018_09_30"

One possible alternative is to mirror the [package search standard specification](https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-search.html#standard-specification) in the CD's coordinates.  
![image](https://github.com/clearlydefined/crawler/assets/92891533/76fb1a72-355f-40d7-ab8a-fe2cce28c319)

The mapping would be as follows: 
```
channel -> provider
subdir -> namespace
name -> name
`${version}-${build}` -> revision 
```
Both version and build may not contain "-" (see https://conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#info-index-json). So using "-" as separator works here.

Additionally, when architecture platform is not specified, should 'noarch' be considered as the default value (https://docs.anaconda.com/anaconda-repository/user-guide/tasks/pkgs/use-noarch-pkgs/)? This likely produces more predictive results.

@elrayle @capfei @jeffwilcox @pombredanne @bduranc I am not particularly knowledgeable in the conda ecosystem, and would very much appreciate other experts' input.

bduranc commented:
I'm okay with both proposals. Qing's is in-line with Conda's own standards, which I assume are applicable to all three channels.

I also assume the _type_ coordinate in both proposals would still be "conda" or "condasource". Also, if there is a GH or GitLab source location present in the channel / package [metadata](https://docs.conda.io/projects/conda-build/en/stable/concepts/generating-index.html), is the intent to populate the same in the definition?

In the past, I believe we've tried our best to keep the coordinate format consistent with the others (for example, when we added Debian/debsrc support back a few years ago, I believe that took some influence from the Maven implementation). It would be best to follow that practice as much as we can here too.

lamarrr commented:
> toolVersion is already used internally by ClearlyDefined as the harvest tool versioning. Using spec.toolVersion will cause conflicts.

I had the impression toolVersion was referring to the tool the package was built with and not of the tooling scanning the licensing info. which I feel would have been less ambiguous if it was at the beginning of the package coordinate.

> Adding /py36 adds complexity to [service APIs](https://api.clearlydefined.io/api-docs/#/). For instance, harvest data api expects
/harvest/{type}/{provider}/{namespace}/{name}/{revision}/{tool}.

Agreed, I'll make it an optional parameter appended to the revision instead

> Concatenating version and architecture with _ does not handle versions that contain _, e.g. "version": "1.30.0_2018_09_30"
Agreed, but I couldn't find a better delimiter to use. I performed a regex search on some of the channels and none of them had that kind of versioning (with '_' in them). it's always semantic versioning (numbers and hyphens only with alpha/beta, https://semver.org/).

The new revision can be `{architecture}--{version}-{build}`.

since architecture can be linux-64.

I feel noarch isn't the right thing. noarch is for platform-agnostic packages which may or may not be present. i.e. 7zip isn't platform-agnostic but is architecture and os dependent so it is not on the noarch list which would lead to fetching it without specifying the architecture to fail.
We presently select randomly from any architecture the package is available on (just as is done on the debian fetcher) which I feel is a much more reliable method than using noarch by default. It might be better to make the subdir/architecture&os required than using noarch. 

asides, subdir isn't really a namespace, it's just an architecture and os folder grouping of the packages (i.e. linux x64 packages -> /linux-64, windows x64 packages -> /windows-64)

lamarrr commented:
> I also assume the _type_ coordinate in both proposals would still be "conda" or "condasource". Also, if there is a GH or GitLab source location present in the channel / package [metadata](https://docs.conda.io/projects/conda-build/en/stable/concepts/generating-index.html), is the intent also to populate the same in the definition?

I don't get you, what definition?
If it is a condasource type, it is sourced from whatever source url or source git url (git is **NOT** always the source) is provided at the package's channel index.
if it is a conda type (architecture-dependent) it is sourced from conda's server source url

bduranc commented:
> > I also assume the _type_ coordinate in both proposals would still be "conda" or "condasource". Also, if there is a GH or GitLab source location present in the channel / package [metadata](https://docs.conda.io/projects/conda-build/en/stable/concepts/generating-index.html), is the intent also to populate the same in the definition?
> 
> I don't get you, what definition? If it is a condasource type, it is sourced from whatever source url or source git url (git is **NOT** always the source) is provided at the package's channel index. if it is a conda type (architecture-dependent) it is sourced from conda's server source url

Basically, a "definition" == component in ClearlyDefined. 

Using Maven as an example:
https://clearlydefined.io/definitions/maven/mavencentral/com.googlecode.openbox/maventools/2.0.1

and it's corresponding "source" definition (Maven sourcearchive): https://clearlydefined.io/definitions/sourcearchive/mavencentral/com.googlecode.openbox/maventools/2.0.1

Or another example that has a GitHub repo maintained as it's source location field instead of the Maven sourcearchive: https://clearlydefined.io/definitions/maven/mavencentral/io.eliez/mavenJava/2.0.1

lamarrr commented:
I have now changed the delimiters and coordinate specification to:

```
{type: conda|condasource}/{provider: anaconda-main|anaconda-r|conda-forge}/-/{package name}/[{archictecture | _}:][{version | _}]-[{build version | _}]/[{tool version}]
```

```
conda/conda-forge/-/numpy/linux-aarch64:1.13.0-py36/ - complete coordinate
conda/conda-forge/-/numpy/-py36/ -- any version with build hash py36*
conda/conda-forge/-/numpy/1.13.0-py36/ -- version with build hash
conda/conda-forge/-/numpy/linux-aarch64:_-py36/ -- architecture and build hash
conda/conda-forge/-/numpy/linux-aarch64:1.13.0/ -- architecture and version
conda/conda-forge/-/numpy/ - any
conda/conda-forge/-/numpy/_:_-_ - any
```

lamarrr commented:
> I have now changed the delimiters and coordinate specification to:
> 
> ```
> {type: conda|condasource}/{provider: anaconda-main|anaconda-r|conda-forge}/-/{package name}/[{archictecture | _}:][{version | _}]-[{build version | _}]/[{tool version}]
> ```
> 
> ```
> conda/conda-forge/-/numpy/linux-aarch64:1.13.0-py36/ - complete coordinate
> conda/conda-forge/-/numpy/-py36/ -- any version with build hash py36*
> conda/conda-forge/-/numpy/1.13.0-py36/ -- version with build hash
> conda/conda-forge/-/numpy/linux-aarch64:_-py36/ -- architecture and build hash
> conda/conda-forge/-/numpy/linux-aarch64:1.13.0/ -- architecture and version
> conda/conda-forge/-/numpy/ - any
> conda/conda-forge/-/numpy/_:_-_ - any
> ```

Seems the file indexer encodes the coordinates into files directly from the coordinate spec, meaning paths separated with `:` might not work. I have changed the spec to this:


```
{type: conda|condasource}/{provider: anaconda-main|anaconda-r|conda-forge}/-/{package name}/[{archictecture | _}--][{version | _}]-[{build version | _}]/[{tool version}]
```

```
conda/conda-forge/-/numpy/linux-aarch64--1.13.0-py36/ - complete coordinate
conda/conda-forge/-/numpy/-py36/ -- any version with build hash py36*
conda/conda-forge/-/numpy/1.13.0-py36/ -- version with build hash
conda/conda-forge/-/numpy/linux-aarch64--_-py36/ -- architecture and build hash
conda/conda-forge/-/numpy/linux-aarch64--1.13.0/ -- architecture and version
conda/conda-forge/-/numpy/ - any
conda/conda-forge/-/numpy/_--_-_ - any
```

qtomlinson commented on providers/process/top.js:
```diff
@@ -146,6 +153,59 @@ class TopProcessor extends AbstractProcessor {
+      `Conda top - coordinates: ${packagesCoordinates.length}, start: ${start}, end: ${end}, sliced: ${slicedCoordinates.length}`
+    )
+
+    await request.queueRequests(slicedCoordinates.map(coord => new Request('package', coord)))
```
Assuming condasrc is for source files, 
-the corresponding request for condasrc coordinates would be new Request('source', coord)
-potentially, we can just process the top binary packages from repo data.  Each binary harvest will auto trigger source harvest if present.

qtomlinson reviewed (commented):
Thanks for making the changes as discussed!  
I have reservations on treating condasrc as binary packages in recent commits(see email discussion).  Suggestions on condaFetch, condaExtract, and top below.

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+    return spec && this.channels[spec.provider]
+  }
+
+  //      {type: conda|condasrc}/{provider: anaconda-main|anaconda-r|conda-forge}/{architecture|-}/{package name}/[{version | _}]-[{build version | _}]/[{tool version}]
```
that's just the full coordinate as seen in the clearlydefined spec. for the clearlydefined tool, not conda's tool/build hash. but yeah, i'll remove it

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+      architecture = packageChannelData.subdirs.includes('noarch') ? 'noarch' : packageChannelData.subdirs[0]
+      this.logger.info(`No binary architecture specified for ${spec.name}, using architecture: ${architecture}`)
+    }
+
```
yes, it is

lamarrr replied:
it already does that (fetch the latest one if version is not provided for both conda and condasrc)

qtomlinson replied:
hm... it fetches the one with highest build, not latest version.  See a.packageData.build < b.packageData.build in _matchPackage.

lamarrr replied:
the latest version is contained in packageChannelData.version, and the latest build of the version is obtained by sorting the build id

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+      location: dir.name,
+      registryData: { 'channelData': packageChannelData, downloadUrl },
+      releaseDate: new Date(packageChannelData.timestamp).toUTCString(),
+      declaredLicenses: packageChannelData.license,
```
yes, that's the format the service expects it in

qtomlinson replied:
"releaseDate": "Fri, 11 Jun 2021 02:25:26 GMT" is in the harvested file.

qtomlinson replied:
https://github.com/clearlydefined/service/blob/4ae722e9eaad2bada374df687c0a5dad9e1efe18/lib/utils.js#L116 tries to parse a date to a utc zoned date time.  Line 117 parses the string using ISO format first.   Release dates by most fetches are ISO format.

lamarrr replied:
ok, done

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+      return request.markSkip(`Unrecognized conda provider: ${spec.provider}, must be either of: ${Object.keys(this.channels)}`)
+    }
+    const channelData = await this.getChannelData(this.channels[spec.provider], spec.provider)
+
```
the linter didn't catch that

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+
+    // unless otherwise specified, we fetch the architecture package
+    if (spec.type !== 'conda' && packageChannelData.subdirs.length === 0) {
+      return request.markSkip('No architecture build in package channel data')
```
that's a mistake, should have been === 'conda'

qtomlinson replied:
Line 52 - 60 related with type === conda, can be moved to _downloadCondaPackage, making the handle function simpler to read.

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+
+    if (!repoData) return this.markSkip(request)
+
+    let packageRepoEntries = []
```
ok

lamarrr commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,277 @@
+  }
+
+  async _downloadCondaSourcePackage(spec, request, version, packageChannelData) {
+    if (version && version !== '_' && packageChannelData.version !== version) {
```
'_'  for version is necessary as we need both version and build in the revision, i.e. '{version}-{build}'
which we can't replace version with '-', otherwise we'd have bogus like '--py36'  which would be difficult to parse compared to '_-py36'

qtomlinson replied:
can potentially use empty string when version is not provided here, e.g. `-py36` for `{version}-{build}`.  After splitting using delimiter `-`, empty string for version and py36 for build.  This avoids the introduction of `_` as none for version, which can be confusing as `-` is used as none in architecture.

qtomlinson commented on providers/process/condaExtract.js:
```diff
@@ -0,0 +1,59 @@
+      const spec = this.toSpec(request)
+      const { releaseDate, registryData, declaredLicenses } = request.document
+      request.document = merge(this.clone(request.document), { releaseDate, registryData, declaredLicenses })
+      let sourceCandidates = [
```
To be consistent with the majority of the xxExtract classes, pls refactor line 27 - 47 to _discoverSource function

qtomlinson commented on providers/process/condaExtract.js:
```diff
@@ -0,0 +1,59 @@
+        sourceInfo = SourceSpec.fromObject(spec)
+        sourceInfo.type = 'condasrc'
+        sourceInfo.namespace = null
+        sourceInfo.version = null
```
version is still needed here, the intent here is to discover the source files for the same version as the harvested binary package.  It is understood that package channel data may not have the same version.

qtomlinson commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,202 @@
+      this.logger.info(`No binary architecture specified for ${spec.name}, using architecture: ${architecture}`)
+    }
+    let repoData = undefined
+    if (!(packageChannelData.subdirs.find(x => x === architecture))) {
```
nit, packageChannelData.subdirs[architecture]

lamarrr replied:
it's not an hashmap, it's a list of strings

qtomlinson commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,202 @@
+      return request.markSkip(`failed to fetch and parse repodata json file for channel ${spec.provider} in architecture ${architecture}`)
+    }
+    let packageRepoEntries = this._matchPackage(spec.name, version, buildVersion, repoData)
+    if (packageRepoEntries.length == 0) {
```
===

lamarrr commented on providers/process/condaSrcExtract.js:
```diff
@@ -0,0 +1,25 @@
+const { merge } = require('lodash')
+
+class CondaSrcExtract extends AbstractClearlyDefinedProcessor {
+  constructor(options) {
```
the constructor is necessary as it contains the logger, and the basehandler needs it for logging. otherwise the crawler won't even start up

qtomlinson commented on providers/fetch/condaFetch.js:
```diff
@@ -0,0 +1,203 @@
+    }
+    packageRepoEntries.sort((a, b) => b.packageData.build.localeCompare(a.packageData.build))
+    packageRepoEntries.sort((a, b) => b.packageData.version.localeCompare(a.packageData.version))
+    return packageRepoEntries
```
b.packageData.build.localeCompare(a.packageData.build) puts lower build string first.  Is that the intent?
hm... Using the version string comparison to pick the latest version will run into problems.  For example, in versions, `'2018.12'` < `'2023.09'` <  `'5.0.0'`, `'2023.09'` is the latest version.  The current logic as it stands will not pick `'2023.09'` .  Can be a good unit test case here for `_matchPackage`.
One possible solution is to sort on `packageData.timestamp`, with highest (latest) coming first.  
Another possibility is to set version = version || packageChannelData.version in handle() before branching into processing source and package download.  This way puts the handling of version fallback in one central spot irrespective whether source or package is being fetched.

lamarrr replied:
b.compare(a) picks the greatest string via lexicographical comparison, a.compare(a) picks the smallest.
those are different versioning schemes, it wouldn't make sense for a single project to use different versioning schemes for a single project. yeah, I'll use the timestamp instead.
setting version to packageChannelData if none is specified won't work as the version is not always present.

lamarrr replied:
timestamp is for when it was last modified

lamarrr replied:
I've changed it to a priority-based selection, first prefer the highest version, then prefer the highest timestamp, then prefer the highest build

qtomlinson replied:
yes. you are right about build string comparison.
For latest version, using simple version string comparison will run in issue: lexically '5.0.0' > '10.0.0'. Example versions, '2018.12', '2023.09' and '5.0.0' are from repodata.json for anaconda_main/linux-64/_anaconda_depends, which may be a good test case. The current logic picks _anaconda_depends-5.3.1-py37_1.tar.bz2, while the latest version is 2023.09.

lamarrr replied:
I've changed to using timestamps instead

qtomlinson reviewed (approved):
Thanks for your contribution and incorporating the comments!

capfei commented:
I don't have much knowledge in the crawler and conda but was able to get it running locally. This looks good to me and fine with the naming convention.

RazaAli99 commented:
Hi
